pkg/server/handler.go

```diff
@@ -253,6 +256,86 @@ func WithRequestIdentity(handler http.Handler) http.Handler {
 	})
 }
 
+var (
+	// impersonationGroups is a set of groups that are not allowed to impersonate
+	// if the user does not have a specific group or lower ID (lower = higher permissions).
+	impersonationGroups = map[string]int{
+		authorizationbootstrap.SystemMastersGroup:  0,
+		authorizationbootstrap.SystemKcpAdminGroup: 1,
+	}
+)
+
+// WithImpersonationGatekeeper checks the request for impersonations and validates them,
+// if they are valid. If they are not, will return a 403.
+// We check for impersonation in the request headers, early to avoid it being propagated to
+// the backend services.
+func WithImpersonationGatekeeper(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// Impersonation check is only done when impersonation is requested.
+		// And impersonations is only allowed for the users, who have metadata in the ctx.
+		impersonationUser := req.Header.Get(authenticationapi.ImpersonateUserHeader)
+		if len(impersonationUser) != 0 {
+			requestor, exists := request.UserFrom(req.Context())
+			if !exists {
+				responsewriters.ErrorNegotiated(
+					apierrors.NewForbidden(schema.GroupResource{}, "", fmt.Errorf("impersonation is not allowed for the requestor")),
+					errorCodecs, schema.GroupVersion{}, w, req)
+			}
+
+			if checkImpersonation(requestor.GetGroups(), req.Header[authenticationapi.ImpersonateGroupHeader]) {
+				handler.ServeHTTP(w, req)
+				return
+			}
+		}
+
+		responsewriters.ErrorNegotiated(
+			apierrors.NewForbidden(schema.GroupResource{}, "", fmt.Errorf("impersonation is not allowed for the requestor")),
+			errorCodecs, schema.GroupVersion{}, w, req)
+	})
+}
+
+func checkImpersonation(userGroups []string, requestedGroups []string) bool {
```

Comment:
```suggestion
func validImpersonation(userGroups []string, requestedGroups []string) bool {
```

Reply:
not sure I can follow here with the levels.

What is the logic? Do we have multiple levels of being privileged?

Reply:
yes. `system:masters` allows to impersonate `system:kcp:admin` but not other way around. and `system:kcp:admin` can impersonate anything bellow too. 

overkill? :D

Reply:
We could drop the levels, yes:

- `system:masters` can do everything.
- Nobody else can impersonate `system:masters`, nor `system:kcp:admin`.

Reply:
Got ya, overkill it is :)